Tests for Query + improved error messages + architecture cleanup

ekinanp opened the pull request.

ekinanp committed:
Add tests for Query

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Improve unmarshal error messages

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Refactor expression tests to remove uses of predicate package

This way, we can reuse the expression package in the predicate package.
This will be necessary to properly implement the meta primary

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Remove the *InDomain methods

The same thing could be achieved with proper documentation. E.g. We
could encourage people to _not_ negate primaries of the form
[<primary_name>, <predicate>] as ["NOT", [<primary_name>, <predicate>]]
but instead write it as [<primary_name>, ["NOT", <predicate>]]. So for
example, something like ["NOT", ["path", ["glob", "foo"]]] could be
written as ["path", ["NOT", ["glob", "foo"]]]. The latter still negates
the path _without_ including the start path (which the former would do).

Similarly for value predicates, something like ["NOT", ["number", [">",
"5"]]] could be written as ["number", ["NOT", [">", "5"]]]. This would
still negate the predicate while also returning false for mis-typed
values.

Hence, the *InDomain methods are not necessary. This considerably
simplifies the code, as will be seen in later commits.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Don't reduce expressions

This was originally needed because of the *InDomain methods. Since those
have been removed, we no longer need to reduce the expressions.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented on api/rql/internal/predicate/expression/not.go:
```diff
@@ -44,32 +44,16 @@ func (n *not) Unmarshal(input interface{}) error {
+	return !n.p.(rql.EntryPredicate).EvalEntry(e)
 }
 
 func (n *not) EvalEntrySchema(s *rql.EntrySchema) bool {
```
This is still a bit of an issue b/c something like `["NOT", ["path", <string_p>]]` would cause the SchemaP to return false. Again, we could document that negation only makes sense for the BooleanPredicate, and to not negate anything else. Could also enforce this in the AST by defining PEs as being made up of only AND/OR, and NPEs (negatable PE) as being made up of AND/OR/NOT. Then the RQL could be something like

```
Query := PE Primary
...
NamePrimary => ["name", NPE StringPredicate]
...
```

Not sure which approach is best. Keeping everything as PEs results in simpler code, but some (possibly) annoying edge cases/debugging. SO maybe NPEs are a good idea.

MikaelSmith replied:
Annoying edge cases/debugging are also code, so probably debatable whether it's simpler. NPE sounds good for clarity.

Will `["NOT", ["path", <string_p>]]` now cause an error?

ekinanp replied:
Yes it will. But `["path", [NOT, <string_p>]]` wont. And that kind of makes sense, you are querying the "path" field and that has a predicate assigned. So NPE seems like the right direction.

ekinanp replied:
I'm also wondering whether the `true/false` primaries are necessary.. Can't really think of a good reason to include them. I kind of wonder if they were there b/c they made it easy to test the operators back in the old days, where writing unit tests was a lot harder.

ekinanp replied:
`true`/`false` definitely make sense as value predicates though.

ekinanp replied:
So another thing that's nice about this is negation matches user intent. So something like `["NOT", ["object", [["key", "foo"], p]]` reads "return true if NOT an object w/ key foo whose value satisfies p", while `["object", [["key", "foo"], ["NOT, p]]]` reads "return true if obj w/ key foo whose value satisfies NOT(p)".

Similarly, `["NOT", ["number, p]]` vs. `["number", [NOT, p]]` -- former reads "return true if _not_ number" satisfying p, other is "return true if number satisying NOT p".

`wash find` will probably still need its custom negation semantics, but that's b/c we're assuming user intent is always to trickle-down the NOT. Different front ends can change their parsing rules to accommodate different user intent, but the RQL's user intent is still generic.

MikaelSmith replied:
I think reducing the number of ways to express the same thing makes sense.

ekinanp replied:
Meh, may also be good enough to keep `wash find` symmetric with RQL once that refactor's done.

MikaelSmith commented on api/rql/internal/primary/cname_test.go:
```diff
@@ -77,7 +67,7 @@ func (s *CNameTestSuite) TestExpression_AtomAndNot() {
 	s.EEFTC(expr, e)
 
-	s.EESTTC(expr, schema)
+	//s.EESTTC(expr, schema)
```
Why is this commented out?

MikaelSmith commented on api/rql/internal/primary/name_test.go:
```diff
@@ -77,7 +67,7 @@ func (s *NameTestSuite) TestExpression_AtomAndNot() {
 	s.EEFTC(expr, e)
 
-	s.EESTTC(expr, schema)
+	//s.EESTTC(expr, schema)
```
Commented out?

MikaelSmith commented on api/rql/internal/primary/path_test.go:
```diff
@@ -72,14 +64,12 @@ func (s *PathTestSuite) TestExpression_AtomAndNot() {
 	s.EEFTC(expr, e)
 
-	s.EESTTC(expr, schema)
+	//s.EESTTC(expr, schema)
```
Commented?

MikaelSmith commented on api/rql/internal/primary/timeAttr_test.go:
```diff
@@ -82,5 +72,5 @@ func (s *TimeAttrTestSuite) TestExpression_AtomAndNot() {
 	s.EEFTC(expr, e)
 
-	s.EESTTC(expr, schema)
+	//s.EESTTC(expr, schema)
```
Commented?

ekinanp replied:
Ah yeah, I wanted to hold off deleting this until NPE/PE was resolved. Since NPE's are OK, I'll remove this commented stuff.

ekinanp replied:
Updated.

ekinanp committed:
Remove commented out s.EESTTC

Since we're going with NPEs, EESTTC no longer needs to be tested at the
specified pieces of code.

Signed-off-by: Enis Inan <enis.inan@puppet.com>
